compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -449,6 +453,53 @@ class TreeChecker extends Phase with SymTransformer {
       tree
     }
   }
+
+  /**
+    * Checks that `New` nodes are always wrapped inside `Select` nodes.
+    */
+  private object WrappedNewChecker {
+
+    // Current state of the search. We return as soon as we see the first improperly wrapped `New`.
+    sealed trait State
+    case class NotFound(parent: Option[tpd.Tree]) extends State
+    case class Found(parent: Option[tpd.Tree], child: tpd.New) extends State
```

Comment:
When I invoke the accumulator on the root of the tree, there's no parent node. But I guess I could also pass an empty Thicket?